crates/block-producer/src/mempool/mod.rs

```diff
@@ -0,0 +1,275 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet, VecDeque},
+    fmt::Display,
+    ops::Sub,
+    sync::Arc,
+};
+
+use batch_graph::BatchGraph;
+use inflight_state::{InflightState, StateDiff};
+use miden_objects::{
+    accounts::AccountId,
+    notes::{NoteId, Nullifier},
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+use miden_tx::{utils::collections::KvMap, TransactionVerifierError};
+use transaction_graph::TransactionGraph;
+
+use crate::{
+    errors::AddTransactionErrorRework,
+    store::{TransactionInputs, TxInputsError},
+};
+
+mod batch_graph;
+mod inflight_state;
+mod transaction_graph;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchJobId(u64);
+
+impl Display for BatchJobId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BatchJobId {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
```

Comment:
This probably belongs in `base`? Something else I've been itching to do is new-type all of the different fields in the block header for example. Having all of them be `Digest`/`u32` isn't great.

Reply:
Yes, I think moving this to `miden-base` would be a good idea. Would you like to make a PR there?

For the other fields - also agreed, but maybe we do these incrementally over time?

Reply:
Will do. And incrementally is good; no rush :) One can also over do it.